homeassistant/components/tuya/__init__.py

```diff
@@ -0,0 +1,233 @@
+            api.login, entry.data[CONF_USERNAME], entry.data[CONF_PASSWORD]
+        )
+    else:
+        response = await hass.async_add_executor_job(
+            api.login,
+            entry.data[CONF_USERNAME],
+            entry.data[CONF_PASSWORD],
+            entry.data[CONF_COUNTRY_CODE],
+            entry.data[CONF_APP_TYPE],
+        )
+
+    if response.get("success", False) is False:
+        _LOGGER.error("Tuya login error response: %s", response)
+        return False
+
+    tuya_mq = TuyaOpenMQ(api)
+    tuya_mq.start()
+
+    device_manager = TuyaDeviceManager(api, tuya_mq)
+
+    # Get device list
+    home_manager = TuyaHomeManager(api, tuya_mq, device_manager)
+    await hass.async_add_executor_job(home_manager.update_device_cache)
+    hass.data[DOMAIN][TUYA_HOME_MANAGER] = home_manager
+
+    listener = DeviceListener(hass)
+    hass.data[DOMAIN][TUYA_MQTT_LISTENER] = listener
+    device_manager.add_device_listener(listener)
+    hass.data[DOMAIN][TUYA_DEVICE_MANAGER] = device_manager
+
+    # Clean up device entities
+    await cleanup_device_registry(hass)
+
+    _LOGGER.debug("init support type->%s", TUYA_SUPPORT_HA_TYPE)
+
+    for platform in TUYA_SUPPORT_HA_TYPE:
+        _LOGGER.debug("tuya async platform-->%s", platform)
+        hass.async_create_task(
+            hass.config_entries.async_forward_entry_setup(entry, platform)
+        )
+
+    return True
+
+
+async def cleanup_device_registry(hass: HomeAssistant) -> None:
+    """Remove deleted device registry entry if there are no remaining entities."""
+
+    __device_registry = device_registry.async_get(hass)
+    device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+
+    for dev_id, device_entity in list(__device_registry.devices.items()):
+        for item in device_entity.identifiers:
+            if DOMAIN == item[0] and item[1] not in device_manager.device_map:
+                __device_registry.async_remove_device(dev_id)
+                break
+
+
+@callback
+def async_remove_hass_device(hass: HomeAssistant, device_id: str) -> None:
+    """Remove device from hass cache."""
+    __device_registry = device_registry.async_get(hass)
+    for entity in list(__device_registry.devices.values()):
+        if device_id in list(entity.identifiers)[0]:
+            __device_registry.async_remove_device(entity.id)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unloading the Tuya platforms."""
+    _LOGGER.debug("integration unload")
+    unload = await hass.config_entries.async_unload_platforms(
+        entry, TUYA_SUPPORT_HA_TYPE
+    )
+    if unload:
+        __device_manager = hass.data[DOMAIN][TUYA_DEVICE_MANAGER]
+        __device_manager.mq.stop()
+        __device_manager.remove_device_listener(hass.data[DOMAIN][TUYA_MQTT_LISTENER])
+
+        hass.data.pop(DOMAIN)
+
+    return unload
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
```

Comment:
Please move `async_setup_entry` up first before `_init_tuya_sdk`. That is easier to read and standard.